### PR TITLE
Websocket review

### PR DESCRIPTION
**Pull request by gregw**
Just creating a PR to use github tools to discuss some alternate implementations

### CODE REVIEW

File: jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/FrameFlusher.java [hunk 1]
```diff
@@ -293,10 +295,10 @@ public String toString()
     private final int bufferSize;
     private final Generator generator;
     private final int maxGather;
-    private final Object lock = new Object();
+    private final Locker lock = new Locker();
```
Comment (gregw) [inline jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/FrameFlusher.java]: I'm using a Locker... not for any real reason other than we use it elsewhere. Synchronized is probably identical

File: jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/FrameFlusher.java [hunk 2]
```diff
@@ -293,10 +295,10 @@ public String toString()
...
     private final Deque<FrameEntry> queue = new ArrayDeque<>();
     private final Flusher flusher;
-    private final AtomicBoolean closed = new AtomicBoolean();
+    private boolean closed = false;
```
Comment (gregw) [inline jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/FrameFlusher.java]: Mixing Atomic state variables and locked blocks of code is pretty much an anti pattern.  Use one or the other and never both.

File: jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/FrameFlusher.java [hunk 3]
```diff
@@ -311,20 +313,26 @@ public FrameFlusher(ByteBufferPool bufferPool, Generator generator, EndPoint end
...
     public void close()
     {
-        if (closed.compareAndSet(false,true))
+        List<FrameEntry> entries = null;
+        
+        try(Locker.Lock l = lock.lock())
         {
-            LOG.debug("{} closing {}",this);
-            EOFException eof = new EOFException("Connection has been closed locally");
-            flusher.failed(eof);
```
Comment (gregw) [inline jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/FrameFlusher.java]: This call would call the callbacks of any entries that had been partially processed inside the locked section of code.  This is asking for deadlocks etc.

File: jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/FrameFlusher.java [hunk 4]
```diff
@@ -311,20 +313,26 @@ public FrameFlusher(ByteBufferPool bufferPool, Generator generator, EndPoint end
...
+
+                entries = new ArrayList<>();
                 entries.addAll(queue);
                 queue.clear();
             }
-            // Notify outside sync block.
+        }
+
+        // Notify outside sync block.
+        if (entries != null)
```
Comment (gregw) [inline jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/FrameFlusher.java]: Do all the notifications here - of entries in the Flusher plus entries in the Q

File: jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/FrameFlusher.java [hunk 5]
```diff
@@ -334,52 +342,63 @@ public void close()
 
     public void enqueue(Frame frame, WriteCallback callback, BatchMode batchMode)
     {
-        if (closed.get())
```
Comment (gregw) [inline jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/FrameFlusher.java]: This looks like a race. Testing atomic variable and then conditionally grabbing a lock to act.

File: jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/FrameFlusher.java [hunk 6]
```diff
@@ -334,52 +342,63 @@ public void close()
...
-        {
-            notifyCallbackFailure(callback,new EOFException("Connection has been closed locally"));
-            return;
-        }
-        if (flusher.isFailed())
```
Comment (gregw) [inline jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/FrameFlusher.java]: Ditto possible race

**Comment by joakime** [comment]
Good stuff.

I'm going to merge this into the issue-207 branch and work with it from there.

**Comment by joakime** [comment]
Merged into `issue-207` branch on commits ca2423985c63a52258ff1e6889d37aeb2b1e1b22 and f9e2f5f96e1b8b4a64c64b784c4607d79e75efa8

**Pull request closed without merging**